backlog_controller.hh

```diff
@@ -126,22 +126,32 @@ public:
 };
 
 class compaction_controller : public backlog_controller {
+    static constexpr float base_normalization_factor = 30.0f;
```

Comment:
From the description I expected the patch to introduce shares = std::min(shares, max_shares), but I see something a lot more complicated. Is it doing something different?

Reply:
If I do something like std::min(shares, max_shares), it means the slope of linear function is the same, and we're capping the max (possible) output.

But what if we want to allow controller to output more than 1k, which is the constant max output of last point?

Then the alternative is to change the last point's output (y2) to the configured max shares.

But then we're changing the slope, as slope is `m = (y2 – y1) / (x2 – x1)`

x2 is the the normalization factor (the last point's input), so by playing with slope formula, we can find a x2 which yield the same slope for any max shares configuration.

See the effects of this approach:

```
            Max shares
Backlog     500         1000        2000
         |=================================
1        |  83.3333     83.3333     83.3333
2        |  115.789     115.789     115.789
4        |  178.947     178.947     178.947
8        |  305.263     305.263     305.263
16       |  500         557.895     557.895
32       |  500         1000        1063.16
64       |  500         1000        2000
```

so we're essentially doing std::min(shares, max_shares), but without the restriction of 1k limit.

I added some sanity tests, on different settings, to make sure controller remains sane.

Reply:
Sorry for losing the thread. I think we should default the cap at 1000 as it is now, and allow the computation to extrapolate indefinitely. So you can change the slope, or change the cap, and the defaults result in unchanged behavior.